File: Lua/Lexer/token.lua
---@class Token
---@field literal string?
---@field type string?
local Token = {
    literal = nil,
    type = nil,
}

---@enum TokenType
local TokenType = {
    EOF = "EOF",
    ILLEGAL = "ILLEGAL",

    -- Core
    LET = "LET",
    CONST = "CONST",
    STRUCT = "STRUCT",
    FUNCTION = "FUNCTION",
    IF = "IF",
    ELSE = "ELSE",
    SWITCH = "SWITCH",
    CASE = "CASE",
    BREAK = "BREAK",
    RETURN = "RETURN",
    WHILE = "WHILE",
    FOR = "FOR",
    AND = "AND",
    OR = "OR",
    IN = "IN",

    -- Types
    IDENTIFIER = "IDENTIFIER",
    NUMBER = "NUMBER",
    FLOAT = "FLOAT",
    BOOLEAN = "BOOLEAN",
    STRING = "STRING",
    CHAR = "CHAR",

    -- Datatypes
    INT = "INT",
    F32 = "F32",
    BOOL = "BOOL",
    STR = "STR",
    NIL = "NIL",
    VOID = "VOID",

    -- Single Characters
    EXCLAMATION = "EXCLAMATION", -- !
    AT = "AT", -- @
    HASHTAG = "HASHTAG", -- #
    DOLLAR = "DOLLAR", -- $
    PERCENT = "PERCENT", -- %
    CARET = "CARET", -- ^
    AMPERSAND = "AMPERSAND", -- &
    ASTERISK = "ASTERISK", -- *
    LPAREN = "LPAREN", -- (
    RPAREN = "RPAREN", -- )
    MINUS = "MINUS", -- -
    UNDERSCORE = "UNDERSCORE", -- _
    PLUS = "PLUS", -- +
    ASSIGN = "ASSIGN", -- =
    LBRACKET = "LBRACKET", -- [
    RBRACKET = "RBRACKET", -- ]
    LBRACE = "LBRACE", -- {
    RBRACE = "RBRACE", -- }
    SEMICOLON = "SEMICOLON", -- ;
    COLON = "COLON", -- :
    APOSTROPHE = "APOSTROPHE", -- '
    QUOTE = "QUOTE", -- "
    COMMA = "COMMA", -- ,
    PERIOD = "PERIOD", -- .
    LESSTHAN = "LESSTHAN", -- <
    GREATERTHAN = "GREATERTHAN", -- >
    SLASH = "SLASH", -- /
    QUESTION = "QUESTION", -- ?
    BACKSLASH = "BACKSLASH", -- \/
    PIPE = "PIPE", -- |

    -- Double Characters
    EQUAL = "EQUAL", -- ==
    NOT_EQUAL = "NOT_EQUAL", -- !=
    PLUS_EQ = "PLUS_EQ", -- +=
    MINUS_EQ = "MINUS_EQ", -- -=
    MULT_EQ = "MULT_EQ", -- *=
    DIV_EQ = "DIV_EQ", -- /=
    LT_EQUAL = "LT_EQUAL", -- <=
    GT_EQUAL = "GT_EQUAL", -- >=
    INCREMENT = "INCREMENT", -- ++
    DECREMENT = "DECREMENT", -- --
    COMMENT = "COMMENT", -- //
}

---@param type TokenType
---@param literal string
---@return Token
local function newToken(type, literal)
    return {
        type = type,
        literal = literal,
    }
end

---@type {[string]: TokenType}
local IdentifierMap = {
    ["let"] = TokenType.LET,
    ["const"] = TokenType.CONST,
    ["struct"] = TokenType.STRUCT,
    ["fn"] = TokenType.FUNCTION,
    ["if"] = TokenType.IF,
    ["else"] = TokenType.ELSE,
    ["switch"] = TokenType.SWITCH,
    ["case"] = TokenType.CASE,
    ["break"] = TokenType.BREAK,
    ["return"] = TokenType.RETURN,
    ["while"] = TokenType.WHILE,
    ["for"] = TokenType.FOR,
    ["and"] = TokenType.AND,
    ["or"] = TokenType.OR,
    ["in"] = TokenType.IN,

    ["true"] = TokenType.BOOLEAN,
    ["false"] = TokenType.BOOLEAN,

    ["bool"] = TokenType.BOOL,
    ["int"] = TokenType.INT,
    ["f32"] = TokenType.F32,
    ["str"] = TokenType.STR,
    ["nil"] = TokenType.NIL,
    ["void"] = TokenType.VOID,
}

local DoubleTokenMap = {
    ["=="] = TokenType.EQUAL,
    ["!="] = TokenType.NOT_EQUAL,
    ["+="] = TokenType.PLUS_EQ,
    ["-="] = TokenType.MINUS_EQ,
    ["*="] = TokenType.MULT_EQ,
    ["/="] = TokenType.DIV_EQ,
    ["<="] = TokenType.LT_EQUAL,
    [">="] = TokenType.GT_EQUAL,
    ["++"] = TokenType.INCREMENT,
    ["--"] = TokenType.DECREMENT,
    ["//"] = TokenType.COMMENT,
}

return {
    Token = Token,
    TokenType = TokenType,
    IdentifierMap = IdentifierMap,
    DoubleTokenMap = DoubleTokenMap,
    newToken = newToken,
}

File: Rust/Lexer/src/token.rs
#![allow(dead_code)]

use std::{
    collections::HashMap,
    sync::{Mutex, MutexGuard, OnceLock},
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenType {
    Eof,
    Illegal,

    // Core
    Let,
    Const,
    Struct,
    Function,
    If,
    Else,
    Switch,
    Case,
    Break,
    Return,
    While,
    For,
    And,
    Or,
    In,

    // Types
    Identifier,
    Number,
    Float,
    Boolean,
    String,
    Char,

    // Datatypes
    Int,
    F32,
    Bool,
    Str,
    Nil,
    Void,

    // Single Characters
    Exclamation, // !
    At,          // @
    Hashtag,     // #
    Dollar,      // $
    Percent,     // %
    Caret,       // ^
    Ampersand,   // &
    Asterisk,    // *
    Lparen,      // (
    Rparen,      // )
    Minus,       // -
    Underscore,  // _
    Plus,        // +
    Assign,      // =
    Lbracket,    // [
    Rbracket,    // ]
    Lbrace,      // {
    Rbrace,      // }
    Semicolon,   // ;
    Colon,       // :
    Apostrophe,  // '
    Quote,       // "
    Comma,       // ,
    Period,      // .
    Lessthan,    // <
    Greaterthan, // >
    Slash,       // /
    Question,    // ?
    Backslash,   // /
    Pipe,        // |

    // Double Characters
    Equal,     // ==
    NotEqual,  // !=
    PlusEq,    // +=
    MinusEq,   // -=
    MultEq,    // *=
    DivEq,     // /=
    LtEqual,   // <=
    GtEqual,   // >=
    Increment, // ++
    Decrement, // --
    Comment,   // //
}

#[derive(Debug, Clone, Copy)]
pub struct Token<'a> {
    pub literal: &'a str,
    pub token_type: TokenType,
}

impl Token<'_> {
    pub fn new(token_type: TokenType, literal: &str) -> Token {
        Token { literal, token_type }
    }
}

pub fn get_identifier_map() -> MutexGuard<'static, HashMap<&'static str, TokenType>> {
    static IDENTIFIER_MAP: OnceLock<Mutex<HashMap<&str, TokenType>>> = OnceLock::new();
    IDENTIFIER_MAP
        .get_or_init(|| {
            use TokenType::*;
            let mut m = HashMap::new();
            m.insert("let", Let);
            m.insert("const", Const);
            m.insert("struct", Struct);
            m.insert("fn", Function);
            m.insert("if", If);
            m.insert("else", Else);
            m.insert("switch", Switch);
            m.insert("case", Case);
            m.insert("break", Break);
            m.insert("return", Return);
            m.insert("while", While);
            m.insert("for", For);
            m.insert("and", And);
            m.insert("or", Or);
            m.insert("in", In);

            m.insert("true", Boolean);
            m.insert("false", Boolean);

            m.insert("bool", Bool);
            m.insert("int", Int);
            m.insert("f32", F32);
            m.insert("str", Str);
            m.insert("nil", Nil);
            m.insert("void", Void);
            Mutex::new(m)
        })
        .lock()
        .expect("Let's hope the lock isn't poisoned")
}

pub fn get_double_token_map() -> MutexGuard<'static, HashMap<&'static str, TokenType>> {
    static DOUBLE_TOKEN_MAP: OnceLock<Mutex<HashMap<&str, TokenType>>> = OnceLock::new();
    DOUBLE_TOKEN_MAP
        .get_or_init(|| {
            use TokenType::*;
            let mut m = HashMap::new();
            m.insert("==", Equal);
            m.insert("!=", NotEqual);
            m.insert("+=", PlusEq);
            m.insert("-=", MinusEq);
            m.insert("*=", MultEq);
            m.insert("/=", DivEq);
            m.insert("<=", LtEqual);
            m.insert(">=", GtEqual);
            m.insert("++", Increment);
            m.insert("--", Decrement);
            m.insert("//", Comment);
            Mutex::new(m)
        })
        .lock()
        .expect("Let's hope the lock isn't poisoned")
}
